content/chainguard/fips/non-approved-algorithms.md

```diff
@@ -0,0 +1,181 @@
+
+Although SHA1 is currently approved, it is already deprecated by RFCs. NIST is deprecating SHA1 by 2030.  The implementations below are forward-looking and attempt to address access to MD5 today and SHA1 in the future.
+
+SHA1 is available as approved in Chainguard FIPS Provider for OpenSSL versions 3.0.9, 3.1.2 and 3.4.0. It is non-approved starting in version 3.6.0. The below guidance will apply to SHA1 as well likely in 2027.
+
+## Access to unapproved algorithms for non-security purposes
+
+### OpenSSL FIPS and MD5
+
+In Chainguard FIPS images, OpenSSL operates in approved only mode and has the default property query string set to `fips=yes` for all algorithms and services, such as message digests, HMAC, and so on.
+
+MD5 is available as a non-approved service. You can request access to it on opt-in basis using a `?fips=yes` (prefer fips implementation if there is one, and fallback to a non-fips one) or `-fips` (disregard request for a fips implementation, and return any available one) property query strings via the C API or via command line options to calculate message digests. Usage of these digests in higher level algorithms is blocked. For example `openssl dgst` calculation is possible, yet `openssl dsgst -sign, -verify, -hmac` is blocked.
+
+Examples:
+
+```shell
+$ echo chainguard | openssl dgst -md5
+Error setting digest
+801B3D417B7F0000:error:0308010C:digital envelope routines:inner_evp_generic_fetch:unsupported:crypto/evp/evp_fetch.c:375:Global default library context, Algorithm (MD5 : 89), Properties ()
+801B3D417B7F0000:error:03000086:digital envelope routines:evp_md_init_internal:initialization error:crypto/evp/digest.c:271:
+```
+
+```shell
+$ printf chainguard | openssl dgst -propquery '?fips=yes' -md5
+MD5(stdin)= f8d689ee8221617c032b0e71f9c597ac
+```
+
+Note that calculating HMAC is still blocked:
+
+```shell
+$ printf chainguard | openssl dgst -propquery '?fips=yes' -md5 -hmac averylonghmackey
+Error setting context
+80AB9D605C7F0000:error:0308010C:digital envelope routines:inner_evp_generic_fetch:unsupported:crypto/evp/evp_fetch.c:341:FIPS internal library context, Algorithm (md5 : 0), Properties (<null>)
+```
+
+There are some caveats and bypasses, some digital signature algorithms allow signing raw data, or prehashed values. In such cases, one can out of band calculate MD5 hash, padded it according to PKCSv1.5 and RSA2048 modulus size and execute raw RSA signing operation. Such a service is non-approved, and is unable to know that it is abused to sign an MD5 hash instead of SHA256. Please don't do this (your FIPS auditors will make you change it). Higher level one shot EVP APIs typically accept a message to sign, and perform hashing and padding internally and correctly block creating MD5 signatures. This again highlights that FIPS is about consent, one has to try to use FIPS cryptography intentionally.
```

Comment:
```suggestion
There are some caveats and bypasses, some digital signature algorithms allow signing raw data, or prehashed values. In such cases, one can calculate MD5 hash out of band, pad it according to PKCSv1.5 and RSA2048 modulus size and execute raw RSA signing operation. Such a service is non-approved, and it is not possible to know that it is being abused to sign an MD5 hash instead of SHA256. _Please don't do this_ (your FIPS auditors will require that you change it). Higher level one-shot EVP APIs typically accept a message to sign, and perform hashing and padding internally and correctly block creating MD5 signatures. This again highlights that FIPS is about consent, one should use FIPS cryptography intentionally.
```